python/feathub/processors/flink/table_builder/redis_utils.py

```diff
@@ -22,20 +23,117 @@
     expressions as native_flink_expr,
 )
 
+from feathub.common import types
 from feathub.common.exceptions import FeathubException
 from feathub.feature_tables.sinks.redis_sink import RedisSink
-from feathub.feature_tables.sources.redis_source import NAMESPACE_KEYWORD, KEYS_KEYWORD
+from feathub.feature_tables.sources.redis_source import (
+    NAMESPACE_KEYWORD,
+    KEYS_KEYWORD,
+    RedisSource,
+    FEATURE_NAME_KEYWORD,
+    KEY_COLUMN_PREFIX,
+)
+from feathub.feature_views.derived_feature_view import DerivedFeatureView
 from feathub.processors.constants import EVENT_TIME_ATTRIBUTE_NAME
 from feathub.processors.flink.flink_jar_utils import find_jar_lib, add_jar_to_t_env
+from feathub.processors.flink.flink_types_utils import to_flink_schema
 from feathub.processors.flink.table_builder.flink_sql_expr_utils import (
     to_flink_sql_expr,
 )
 from feathub.processors.flink.table_builder.source_sink_utils_common import (
     get_schema_from_table,
 )
+from feathub.table.schema import Schema
 from feathub.table.table_descriptor import TableDescriptor
 
 
+def get_redis_source(feature_desc: TableDescriptor) -> Optional[RedisSource]:
+    """
+    Returns the RedisSource if the feature descriptor represents a RedisSource
+    optionally followed by a series of DerivedFeatureView, which is the valid
+    syntax before a lookup join using the RedisSource. Returns None if the
+    feature descriptor does not match this syntax.
+    """
+    if isinstance(feature_desc, RedisSource):
+        return feature_desc
+    if isinstance(feature_desc, DerivedFeatureView):
+        if isinstance(feature_desc.source, str):
+            raise FeathubException(
+                f"feature descriptor {feature_desc.name} must have been resolved."
+            )
+        return get_redis_source(feature_desc.source)
+    return None
+
+
+def append_redis_key_columns(
```

Comment:
I think it might be better to adopt the concept of physical keys and logical keys in general and renamed it to `append_physical_key_columns`. Please check if this is proper.

Reply:
Sounds good.